Track long lived diagnostics inside helix

pascalkuthe opened the pull request:
Fixes #701

This PR adds a configuration option (`persistant-diagnostic-sources`) that allows helix to ignore some diagnostics in `publishDiagnostic`. This option takes affect if all diagnostics from a single `lsp_type::Diagnostic::source` are identical to the last list of diagnostics received `publishDiagnostic` for this source. In that case (If the `source` is part of the `persistant-diagnostic-sources` list) all these diagnostics are ignored and instead, the "old" diagnostics for that source are maintained.

Helix already automatically updates the position of diagnostics based on edits so by keeping the existing diagnostics around we can use this mapping mechanism of helix if the server doesn't map long-lived diagnostics through changes itself.

`rust-analyzer` is a good example here which has many diagnostics which update instantly, but diagnostics generated by the fly-check (rustc/clippy) can only be updated when a file is saved. That means that rust-analyzer will send a publishDiagnostic notification to update it's native diagnostics. It's very hard (and inefficient) to map diagnostics trough changes server side and therefore RA has no choice but to also resend the fly check diagnostics (otherwise they are removed) which undoes the work helix does to map them.

The real root cause of this problem is that the LSP standard currently doesn't allow the server to only update some diagnostics for a file (for example by adding a `source` attribute to the `publishDiagnostic` notification itself). Therefore, I came up with this client-side workaround. This works very well in practice but is not 100% correct because in theory the diagnostics could have actually moved and incidentally all end up at the same position as previously. However, the workaround here compares the entire diagnostic (including message)so the chances of that occurring are extremely slim. Perhaps an LSP extension can be proposed for this use case in the future to handle this properly.

As some of these diagnostics now stick around longer some limitations in helixes own diagnostic mapping became apparent. In b3145f94bd9b44e7776f3b421c225dc481b4eb03 I therefore improved the handling of these cases (detailed description in commit message). The logic implemented here can also be reused for better inlay-hint tracking in the future (in fact VScodes inlay hint tracking partially inspired the implementation).

poliorcetics commented on languages.toml:
```diff
@@ -11,6 +11,7 @@ auto-format = true
 comment-token = "//"
 language-server = { command = "rust-analyzer" }
 indent = { tab-width = 4, unit = "    " }
+persistant-diagnostic-sources = ["rustc", "clippy"]
```
```suggestion
persistent-diagnostic-sources = ["rustc", "clippy"]
```

pascalkuthe commented:
rebased after mege of #2507. I added a small commit to fix some small inconsistencies around diagnostic sorting (and to make sure its always deterministic) since I was touching that code anyways

dead10ck commented on helix-term/src/application.rs:
```diff
@@ -747,31 +748,63 @@ impl Application {
+                            let text = doc.text().clone();
+
+                            let mut unchaged_diag_sources_ = Vec::new();
+                            if let Some(lang_conf) = &lang_conf {
+                                if let Some(old_diagnostics) =
+                                    self.editor.diagnostics.get(&params.uri)
+                                {
+                                    if !lang_conf.persistent_diagnostic_sources.is_empty() {
+                                        // Sort diagnostics first by severity and then by line numbers.
+                                        // Note: The `lsp::DiagnosticSeverity` enum is already defined in decreasing order
+                                        params
+                                            .diagnostics
+                                            .sort_unstable_by_key(|d| (d.severity, d.range.start));
+                                    }
+                                    for source in &lang_conf.persistent_diagnostic_sources {
+                                        let new_diagnostics = params
+                                            .diagnostics
+                                            .iter()
+                                            .filter(|d| d.source.as_ref() == Some(source));
+                                        let old_diagnostics = old_diagnostics
+                                            .iter()
+                                            .filter(|(d, d_server)| {
+                                                *d_server == server_id
+                                                    && d.source.as_ref() == Some(source)
+                                            })
+                                            .map(|(d, _)| d);
+                                        if new_diagnostics.eq(old_diagnostics) {
+                                            unchaged_diag_sources_.push(source.clone())
+                                        }
+                                    }
+                                }
+                            }
```
I wonder if all this logic could just move into `replace_diagnostics`. As it is, the late function has been changed so that it assumes repeat diagnostics have already been removed, and could potentially allow repeats to be added if this invariant is violated.

pascalkuthe replied:
that's not really possible because we need access to `Editor::diagnostics` (not the same as `Document::diagnostics`)  which would be pretty awkward. I also actually prefer to keep the LSP specific logic like thous out of the document if possible.

pascalkuthe replied:
that said this call site doesn't actually handle this correctly. i think I messed that up in the last rebase by accident. I need to rebase this again anyway so I will take a look at that then

dead10ck commented:
Could you explain why we would want this to be configurable, rather than just doing it for all diagnostics? Especially since we're not adding any defaults.

pascalkuthe commented:
> Could you explain why we would want this to be configurable, rather than just doing it for all diagnostics? Especially since we're not adding any defaults.

The reason we don't do this for all diagnostics too is that we consider diagnostics unchanged if the server sends the same diagnostics (made line/col number, message,..) again. This is not really LSP compliant and can lead to situations where diagnostics are not updated when they should (text was edited and diagnostic is now somewhere else that only incidentilally corresponds to the same positions). This is a tradeoff and only with it for diagnostics that actually benefit from this. Namely diagnostics that case only updated on safe (or some other rare event) like rust-analyer `checkOnSafe` diagnostics.

I also added a default for rust-analyzer so this setting is used by default.

pascalkuthe commented:
I improved the word tracking mapping even further the implementation basically matches vscode in behavior and could be used to fix most of the weirdness surrounding inlay hints  (except the low idle timeout)

the-mikedavis commented on helix-core/src/diagnostic.rs:
```diff
@@ -39,6 +39,10 @@ pub enum DiagnosticTag {
+    // whether this diagnostic ends at the end of(or inside) a word
+    pub ends_at_word: bool,
+    pub starts_at_word: bool,
+    pub zero_width: bool,
```
Originally I thought this could be trivially computed but if I'm understanding correctly, a diagnostic may come in as zero-width and then changes may cause `range.start` and `range.end` to shift, so `range.start == range.end` may not detect this case after changes, right?

pascalkuthe replied:
Exactly these are the properties of the diagnostic when sent from the server. The idea is that non zero width diagnostics are removed if they are empty after mapping. But at that point we don't now if they were *previously* empty

the-mikedavis commented on helix-core/src/syntax.rs:
```diff
@@ -154,6 +154,8 @@ pub struct LanguageConfiguration {
     /// Falling back to the current working directory if none are configured.
     pub workspace_lsp_roots: Option<Vec<PathBuf>>,
+    #[serde(default)]
+    pub persistent_diagnostic_sources: Vec<String>,
```
Could you add a blurb about this to book/src/languages.md?

pascalkuthe replied:
sure sorry forgot about that :smile: I added an entry to the docs

the-mikedavis reviewed (commented):
I think this looks good. I've been using it for a while and I like the assoc changes, especially when stacked with #6417. Just a small comment about the docs

the-mikedavis commented on book/src/languages.md:
```diff
@@ -68,6 +68,7 @@ These configuration keys are available:
 | `formatter`           | The formatter for the language, it will take precedence over the lsp when defined. The formatter must be able to take the original file as input from stdin and write the formatted file to stdout |
 | `text-width`          |  Maximum line length. Used for the `:reflow` command and soft-wrapping if `soft-wrap.wrap-at-text-width` is set, defaults to `editor.text-width`   |
 | `workspace-lsp-roots`     | Directories relative to the workspace root that are treated as LSP roots. Should only be set in `.helix/config.toml`. Overwrites the setting of the same name in `config.toml` if set. |
+| `persistent_diagnostic_sources` | An array of LSP diagnostic sources assumed unchanged when the language server resends the same set of diagnostics. Helix can track the position for these diagnostics internally instead. Useful for diagnostics that are recomputed on save.
```
```suggestion
| `persistent-diagnostic-sources` | An array of LSP diagnostic sources assumed unchanged when the language server resends the same set of diagnostics. Helix can track the position for these diagnostics internally instead. Useful for diagnostics that are recomputed on save.
```

`#[serde(rename_all = "kebab-case")]` :P

pascalkuthe replied:
ah damm copy paste got me :D

David-Else commented:
@pascalkuthe This PR has caused immediate problems. Warnings are not shown on screen unless I first save a file. They are shown in the picker immediately. They were shown on the screen immediately before this change.

If you want an example:

1. clone and switch to this branch https://github.com/David-Else/rusty-zombie/tree/crossterm
2. open main, open workspace diagnostics and see the 2 warnings.
3. go to the warnings and see nothing is displayed on the screen
4. save the file and they appear, you will then see:

![Screenshot from 2024-01-06 10-41-07](https://github.com/helix-editor/helix/assets/12832280/27797b9f-e6e1-482d-8cd0-bc2b396e9870)

pascalkuthe commented:
this is not caused by this PR and not a new problem either, see https://github.com/helix-editor/helix/pull/8873

David-Else commented:
> this is not caused by this PR and not a new problem either, see #8873

OK, but it reliably happens every time with this PR, and I have never seen the problem before. Hope to see that fix merged asap :)
